source/utilities.cc

```diff
@@ -745,3 +745,65 @@ namespace aspect
     template std_cxx1x::array<double,3> spherical_coordinates<3>(const Point<3> &position);
   }
 }
+
+namespace dealii
+{
+  namespace Utilities
+  {
+    namespace MPI
+    {
+      template<int dim>
+      Tensor<1,dim> 
+      reduce_tensor(const Tensor<1,dim> &local, const MPI_Comm &mpi_communicator)
+      {
+        std::vector<double> entries(dim);
+        std::vector<double> global_entries(dim);
+        Tensor<1,dim> global;
```

Comment:
The original C language required all declarations to come before any code. This is historical because it was done that way in all previous languages as well, and because it makes it easy for compilers to determine memory layout before memory accesses. But it's a terrible practice from the perspective of code readability ("what type did this variable used in line 189 of the function have again, and where was it previously used?") and it makes it unnecessarily hard for compilers to optimize ("did the function that is called over here between declaration and use modify the variable or not?"). There is a substantial body of research that suggests that moving declarations as far down as possible is good practice all around. Going beyond this, you will find many places in deal.II and ASPECT where we do things like
```
  void f() {
     int i = 42;
     g(i);
     {       // open block
        int k=21;
        h(k);
     }  // close block; make sure 'k' goes out of scope
    ...some more work on 'i'.
  }
```
So we explicitly limit the lifetime of variables both from above and below. Makes code so much easier to understand.